libraries/browser-client/src/client/browser_client.ts

```diff
@@ -0,0 +1,278 @@
+          dataSAB,
+          debug: this.debug,
+        })
+      } else {
+        // Opera, Chrome Android, Webview Android
+        if (this.debug) {
+          console.log('[SpeechlyClient]', 'can not use SharedArrayBuffer')
+        }
+      }
+
+      this.speechlyNode.port.onmessage = (event: MessageEvent) => {
+        switch (event.data.type) {
+          case 'STATS':
+            console.log('received stats from port')
+            if (event.data.signalEnergy > this.stats.maxSignalEnergy) {
+              this.stats.maxSignalEnergy = event.data.signalEnergy
+            }
+            break
+          case 'DATA':
+            this.handleAudio(event.data.frames)
+            break
+          default:
+        }
+      }
+    } else {
+      if (this.debug) {
+        console.log('[SpeechlyClient]', 'using ScriptProcessorNode')
+      }
+      throw Error('no support yet')
+      // Safari, iOS Safari and Internet Explorer
+      // if (this.isWebkit) {
+      //   // Multiply base buffer size of 4 kB by the resample ratio rounded up to the next power of 2.
+      //   // i.e. for 48 kHz to 16 kHz downsampling, this will be 4096 (base) * 4 = 16384.
+      //   const resampleRatio = this.audioContext.sampleRate / this.sampleRate
+      //   const bufSize = baseBufferSize * Math.pow(2, Math.ceil(Math.log(this.resampleRatio) / Math.log(2)))
+      //   this.audioProcessor = this.audioContext.createScriptProcessor(bufSize, 1, 1)
+      // } else {
+      //   this.audioProcessor = this.audioContext.createScriptProcessor(undefined, 1, 1)
+      // }
+      // this.audioContext.createMediaStreamSource(this.mediaStream).connect(this.audioProcessor)
+      // this.audioProcessor.connect(this.audioContext.destination)
+      // this.audioProcessor.addEventListener(audioProcessEvent, (event: AudioProcessingEvent) => {
+      //   this.handleAudio(event.inputBuffer.getChannelData(0))
+      // })
+    }
+    await this.decoder.setSampleRate(this.audioContext?.sampleRate)
+    this.initialized = true
+  }
+
+  async close(): Promise<void> {
+    await this.detach()
+    this.speechlyNode?.port.close()
+    this.speechlyNode?.disconnect()
+    await this.decoder.close()
+    // Disconnect and stop ScriptProcessorNode
+    // if (this.audioProcessor != null) {
+    //   const proc = this.audioProcessor
+    //   proc.disconnect()
+    // }
+  }
+
+  async attach(mediaStream: MediaStream): Promise<void> {
+    await this.initialize()
+    await this.audioContext?.resume()
+
+    if (!this.speechlyNode) {
+      throw Error('cannot attach to stream, no node')
+    }
+    this.stream = this.audioContext?.createMediaStreamSource(mediaStream)
+    this.stream?.connect(this.speechlyNode)
+  }
+
+  async detach(): Promise<void> {
+    if (this.active) {
+      await this.stop()
+    }
+    if (this.stream) {
+      this.stream.disconnect()
+      this.stream = undefined
+    }
+  }
+
+  async uploadAudioData(audioData: ArrayBuffer): Promise<void> {
```

Comment:
Some of our friends requested for this function to return the context id that is used to stream the samples. Now it's difficult (impossible?) to know to which context a given input ArrayBuffer belongs to if one fires several of these concurrently... Or if that's not a good idea, maybe some other way to achieve said mapping.

Reply:
One cannot stream multiple files to the API concurrently, at least not at the moment, except by creating multiple clients. There is a small problem with this function though, as the events will be fired while awaiting it. So, if you do
```
const contextId = await client.uploadAudioData(buf)
```
the contextId is received only after the callbacks are already called. So one cannot really use this upload thing easily.

Reply:
Maybe not the most elegant approach, but what about adding a callback parameter that gets called as soon as the contextId is known? In this situation the new method signature would be 
`async uploadAudioData(audioData: ArrayBuffer, onContextIdReceived: ContextIdCallback): Promise<void>`

Reply:
I agree, it would make it easier to find out what the new `contextId` will be, before the event handlers start singing... My initial idea was that as the API backend for this function does not allow more than 1 concurrent audioContext to be sending audio, the `await` for the `uploadAudioData` would not hurt that much. The upload is also quite fast, compared to the result waiting, so this function would finish way ahead of the `segmentEnd` event that would follow. 
So, one could keep track of incoming events based on their contextId and resolve them after segment end arrives. This could probably be wrapped in a `Promise` though 🤔

Reply:
Interesting! I can see some scenarios where the developer might want to resolve the promise even before the final segment arrives (for example, if the system is just looking for some words in a known dictionary, the partial segments are enough most of the time). I think that in those scenarios, wrapping the entire process in a Promise would probably waste some unnecessary time.

Reply:
Yes, if a "fast" match is wanted, then something like this should already work:

```ts
class MyMatcher {
  private client: BrowserClient
  private matchWords: [string?] = []
  private resolve?: (v: any) => void

  constructor(client: BrowserClient) {
    this.client = client
    this.client.onTranscript((contextId: string, segmentId: number, word: Word) => {
      if (this.resolve) {
        console.log("matcher matching", "word", word.value, "to", this.matchWords)
        if (this.matchWords.some(v => v?.toUpperCase() === word.value.toUpperCase())) {
          console.log("found a match!")
          this.resolve(true)
        }
      }
    })
    this.client.onSegmentChange((s: Segment) => {
      if (this.resolve && s.isFinal) {
          this.resolve(false)
          this.resolve = undefined
      }
    })
  }

  async upload(audioData: ArrayBuffer, matchWords: [string]): Promise<void> {
    this.matchWords = matchWords
    this.client.uploadAudioData(audioData)
    return new Promise((resolve) => {
      this.resolve = resolve
    })
  }
}
```

so, this code will bypass any need to look after the `contextId` by doing the following:
- it registers a listener for transcripts (words) which matches the word to the expected list
- it registers a listener for segment end, which will handle the non-matching case
- it starts the upload without awaiting it, and returning a Promise that will resolve to `true` immediately if the any words match, or `false` if segment end is seen and no match has been found.

As there can be only one active audioContext at a time, this code will work as the Promise received cannot be resolved while the context is active. So one could run this code in loop safely. 

But, there can be only one... so no concurrency with the audioContext is possible. We are planning to create a batch API for this kind of use in the future, if serial processing is not enough.